### PR TITLE
Eval-  precision recall metrics

### PR DESCRIPTION
**Pull request by cvqluu**
# Features Added
- The eval pipeline now outputs the following into an eval directory, given an input test csv:
    - precision, recall, and un-normalized confusion matrix figures 
    - a csv with each example, predicted label, true label
- Some other small changes
    - Improved documentation on `key_label_map` in the `ArgMatcher` class
    - Examples in the test csv which are found in the training set are excluded

Below is the example output of the model using [this test csv](https://gist.github.com/cvqluu/df0323b68f17bc255d546d4c6865e9fd) and the following settings:
- `threshold = 0.5`
- `certain_threshold = 0.9`
- `n_neighbors = 3`

## Normalized by true class (diagonal is precision)

# TODOs:
- ~~Make a small csv with all "single number" metrics all in one~~
    - ~~Balanced acc~~
    - ~~Average precision/recall~~
- Make a test set that appropriately tests the bot behaviour
    - Include more `_na_` examples
- ~~[Bug] Need to investigate why certain examples are classified correctly in test mode and reddit bot but not in `eval.py` output~~

### CODE REVIEW

**Comment by cvqluu** [comment]
Copy-pasting what is in the README:

# Performance Evaluation

Performance metrics for the bot can be obtained by using `eval.py`. If it has not been done before (or if changes to the knowledge base have been made), the pre-computed embeddings should be repopulated using:

```sh
python argmatcher.py
```

A test data `csv` file can then be evaluated by running something like the following command:

```sh
python eval.py --eval-csv <test_data.csv> --n-neighbors 3 --threshold 0.5 --certain-threshold 0.9
```

This `test_data.csv` should be a two column comma separated file with the headings: `text, label`. An example of such a file can be found [here](https://gist.githubusercontent.com/cvqluu/df0323b68f17bc255d546d4c6865e9fd/raw/a5f601d7a23101afd6f97c23a0798b227921d20c/antiveg_comments.csv).

This script will provide the following outputs in a unique `eval_logs/<eval_run_id>/` folder:

- Results summary: `results.csv`
  - Balanced accuracy
  - Averaged Precision and Recall
    - `"micro", "macro", "weighted"`: See [sklearn-docs](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_fscore_support.html#sklearn.metrics.precision_recall_fscore_support)
- Confusion matrices:
  - `cm_true.png`: Normalized by true label (diagonal is per-class precision)
  - `cm_pred.png`: Normalized by pred label (diagonal is per-class recall)
  - `cm_none.png`: Un-normalized
- Every example evaluated: `raw_results.csv`:
  - Contains predicted label, true label

**Comment by cvqluu** [comment]
Here is an example output folder, compressed:

[Link](https://drive.google.com/file/d/1OP0a3-YbqPMndbRc_QtscbDUn90Kyovw/view?usp=sharing)
